### ISSUE TITLE
How to fit continuous N dimentional data using GMMHMM model ?

### ISSUE THREAD

**Issue by eponraj27392** [hmmlearn/hmmlearn#346]
In parallel to  #343, **here i would like to ask the same question with an example**. I hope I will get some answers for this question.

I have a data stored in a list, where each element is an 2 Dimensional Array like below.,

```
X = [ [100*2],  [100*2], [100*2], [100*2], [100*2], .......................[100*2]]
Length of X is 20
```

**For Example : I have shown below the Real Data for X [0], X[1] .....X[19]**

```
X = [   [[-2.77274745,  0.77281236],            [[ 0.21090103, -0.07930404],         ...................... ........ 
           [10.63208364, 11.43457041],             [ 3.73948668,  1.11798521],
           [12.26971883, 13.00240112],              [-0.36787007,  0.79593116],
           [11.81902239, 10.96139825],              [ 1.42856583, -0.1578503 ],
          [10.30706826, 10.51825042],               [ 2.15433481, -0.66107892],
          [ 4.73646468,  6.97560163],                 .
           .                                                           . 
           .                                                           .
           .                                                           . 
           .                                                           .
           .                                                           .
          [20.62377689, 21.33671219]]             [ 0.06788783,  1.57977437]]
```

** Is this the right way of Fitting my data for GMMHMM Training ** 
Based on the document to fit multiple sequence [here](https://hmmlearn.readthedocs.io/en/latest/tutorial.html#multiple-sequences), I tried to concatenate my list X like below code

```

X_1 = np.empty((0, 2))

for i in range(0, len(X)):
    result = X[i]
    X_1 = np.append(X_1 , result, axis=0)

# Now my X_1 looks like shape of 2000*2 (2000 rows and 2 columns)
# Also, lengths = [100]*20

# Now i perform training by initializing some some parameters on my own

model_ = hmm.GMMHMM(n_components=5,n_mix = 3, n_iter=100, init_params='', covariance_type="full",  verbose=True)

model_.startprob_ = 1*5
model_.transmat_= 5*5
model_.weights_ = 5*3
model_.means_ = 5*3*2
model_.sigma_ = 5*3*2*2

model_.fit(X, lengths)
```

**As a result,** 
**Verbose**
         1      -11189.6415             +nan
         2      -10264.9211        +924.7204
         3      -10559.9932        -295.0720

GMMHMM(algorithm='viterbi', covariance_type='full',...........

**Q1** : Whether the steps followed is right or wrong ? If it is wrongs kindly correct me.

**Q2** : Why it run only` 3 iterations`, Also I check for `model_.monitor_` it shows only 3 iterations instead of 100 iterations and it converges ? 

**Q3 :** I don't believe that,  it converges very soon, since i use the same data set in another HMM program which built using theano, it converges in near nearly around the value 8000 with the same initialization parameters ?

Kindly let me the reason. If you need further information please let me know.

**Comment by BianLina** [comment]
@eponraj27392 
Hi, I have the same problem as you. When I use multi-sequence training, I can only iterate twice, but when I use single sample, I can converge. I don't know what the reason is. If you have solved it, would you answer the question? Thank you!

**Comment by eponraj27392** [comment]
Based on my understanding, hmmlearn library support or fit only discrete
multi sequence values. (discrete HMM with GMM)
But when I try to fit continuous values using HMM - GMM API, it converge
within 2 iterations. that is not true in my case.
Thai is what my understanding. (Even i read it in one of the stack answer,
but i cannot able to find now )

Regards / Yours sincerely

---------------------------------------------------------------

Esakkiponraj Esakkimuthu

e-Mail:   eponraj27392@gmail.com

---------------------------------------------------------------

*$$$ The race has not yet over bcoz I haven't won yet** $$$*

**Comment by BianLina** [comment]
@eponraj27392 
Thank you very much for answering my question！
I use the HMM-Gauss model to train multiple sequences that can converge. What do you mean is that hmmlearns are not suitable for multiple sequences of HMM-GMM models? So how do you solve the problem of HMM-GMM multi-sequence training in the end? I know from a blog that hmmlearn library do not seem to support incremental training.
If it is convenient for you, I would like to thank you very much for your reply again！

**Comment by eponraj27392** [comment]
Hallo,
What do you mean is that hmmlearns are not suitable for multiple sequences
of HMM-GMM models?
For multiple sequence continuous data/real value data/non discrete data may
not be suitable to train HMM with GMM model using hmmlearn library. (I
tried to model, but failed)

how do you solve the problem of HMM-GMM multi-sequence training in the end?
Solution 1 : Using github source code, HMM GMM training created from
scratch using deep learning libraries. So i used that information and
perform the training.
Solution 2 : MATLAB has external toolbox ( BNT for the continuous HMM
and Gaussian
mixture. )
I haven't tried but i received suggestion from my friends and members from
other community.

Hope it helps

Regards / Yours sincerely

---------------------------------------------------------------

Esakkiponraj Esakkimuthu

e-Mail:   eponraj27392@gmail.com

---------------------------------------------------------------

*$$$ The race has not yet over bcoz I haven't won yet** $$$*

**Comment by BianLina** [comment]
Thank you very much for your help! 
Conveniently, I would like to know which deep learning library you use to implement HMM-GMM multi-sequence training?
I'm looking forward your reply agian!

Best wishs!

**Comment by eponraj27392** [comment]
Tensor flow and theano.
Refer course: lazy programmer -  Hidden markov model in udemy


Regards / Yours sincerely

---------------------------------------------------------------

Esakkiponraj Esakkimuthu

Mob   :  +49 15259722973

e-Mail:   eponraj27392@gmail.com

---------------------------------------------------------------

*$$$ The race has not yet over bcoz I haven't won yet** $$$*

**Comment by fcostin** [comment]
The symptoms reported in this issue match symptoms reported by users attempting to fit GMMHMM model using multi-sequence data. See:

- issue https://github.com/hmmlearn/hmmlearn/issues/410 which has a self-contained reproducible example illustrating the problem.
- issue https://github.com/hmmlearn/hmmlearn/issues/377 where three users mention similar symptoms but no reproducible example is given

A patch to GMMHMM fit to support multi-sequence data has been merged (see pull request: https://github.com/hmmlearn/hmmlearn/pull/417 ). The patch appears to resolve the symptom for similar issue https://github.com/hmmlearn/hmmlearn/issues/410 .  It is plausible that the issues reported here may be partially or completely resolved by the patch, but it is not possible to confirm as this issue report lacks a self-contained reproducible example.

**Comment by anntzer** [comment]
@fcostin has kindly provided a fix to fit multiple sequences using GMMHMM, which was previously not supported.  This is now released as hmmlearn 0.2.6.  Please consider upgrading to it, and reopen the issue if your problem is still present.
